### ISSUE TITLE
IPTV capabilities ?

### ISSUE THREAD

**Issue by BMitrovic17** [clsid2/mpc-hc#32]
First of all, thanks for continuing the works with what you can.

Is there any way that you could review/merge @xpc1000's PR (https://github.com/mpc-hc/mpc-hc/pull/145) ?

**Comment by clsid2** [comment]
I don't have the hardware to test it.

If they rebase it on latest code and create a test build for users like yourself to test it, then I am willing to merge it if those tests reveal no regressions.

**Comment by Eagle3386** [comment]
@BMitrovic17 I'm an IPTV and MPC-HC user. If you can provide me with a test build, I'd happily assist in testing for you.

**Comment by adipose** [comment]
@Eagle3386 @BMitrovic17 

I have merged the code and created a test build.  The dialogs work fine, but I have no idea if the actual functional code works, so it will need testing.

https://mega.nz/#!5UwwXSRS!1XtA1BKK4lapA_HOwmKvPbxV3EA_BUbulmcruW8R6Ec

If it works well I can create a PR.

**Comment by Eagle3386** [comment]
> I have merged the code and created a test build. The dialogs work fine, but I have no idea if the actual functional code works, so it will need testing.

Awesome! Will be back at home in about an hour and start testing then.

**Comment by adipose** [comment]
What ever happened with this testing?

**Comment by Eagle3386** [comment]
Good morning!

Yes, I've been testing it, but forgot to answer due to moving into our new house.
Sorry about that.

Test failed as MPC did not start the stream while e. g. VLC did. I'll provide more details as soon as I can, but that might take a couple of days, because my PC didn't move in, yet.

**Comment by adipose** [comment]
I actually got this semi-working with the test build.  I added a couple of manual channels, one of which was https://fluxustv.blogspot.com@bit.ly/FluxusTV-Wallpapers?.mp4 (randomly found on the internet here: https://pastebin.com/raw/ZzGTySZE).

![image](https://user-images.githubusercontent.com/3324395/72962662-cdb0d500-3d69-11ea-8fb5-c4d1e7277543.png)


I was able to hit the "skip forward" button, which played the second channel:

![image](https://user-images.githubusercontent.com/3324395/72962543-7874c380-3d69-11ea-943b-a35fc8edff5b.png)

**Comment by adipose** [comment]
@clsid2 , I believe this was mostly working, but no one who expressed interest has been able to provide much testing.  The branch for this fix is suffering from bitrot.  Some of the changes involve reorganizing property pages, which will become more difficult to merge in the future.

I am not confident this can be merged, but I also don't want to waste the work done.  If someone can test it, maybe we can save the work.  Otherwise, I'm afraid the issue should simply be marked wontfix for lack of interest in the feature.

**Comment by PrplHaz4** [comment]
Here are two good resources for testing IPTV:
https://github.com/notanewbie/LegalStream
https://github.com/iptv-org/iptv (also has EPG sources)

Not sure what would need to be tested from an IPTV standpoint.  I am currently using MPC-HC with an M3U8 playlist for monitoring ip cameras on the big screen (flipping through channels w a remote control) - and things are working pretty good as-is.

**Comment by adipose** [comment]
I think some of the featureset was saving iptv channels similar to the capture channels.  Instead of manually maintaining playlists.  All in all it seemed like some decent code but it's been a while.

**Comment by clsid2** [comment]
Favorites menu could also be used to store such channels.

**Comment by adipose** [comment]
Here's everything that was in the original PR that I ported:

Extending the current interface for using it not only for DVB but also for IPTV.

Commits:
1- Add IPTV capabilities to the existing DVB UI

    Basic IPTV capabilities
    Navigation TV panel: tabbed filters for DVB-T, IPTV and Radio

2- Digital TV: adapting the channel management UI for IPTV:

    Channel Scan is now called Digital TV tools
    UI: new tab for managing IPTV channels
    IPTV: option for adding individual channels
    IPTV: option for importing a m3u list of channels
    New checkbox for removing existing channels when saving. Removes only the channels of the same type (DVB or IPTV)
    Adding channels: new counter for preventing having duplicate prefNumber channels

3- Digital TV: split options from analog capture and new Digital TV options

3.1. General

    Split menu entry into: Open Digital TV / Open Analog Capture. File->Open Device now only refers to analog device
    Split class OpenDeviceData into OpenDeviceAnalogData and OpenDeviceDigitalData
    Channel navigation panel: only enabled tabs visible (depending on Digital TV options). Set initial active tab depending on current channel type.
    Add IGMP membership management: manage IGMP group join/leave for multicast IPTV streaming sources. Supports only IPv4 and IGMPv2. IPv6 and IGMPv3 could be added in the future if needed.
    Add shortcut for Digital TV

3.2. Exception handling:

    error controls (checking devices, etc.) moved from the general entry point to each channel switch operation
    Change current behavior: when error switching to a channel -> show error message in info bar and then stay ready for switching to another channel
    If the IPTV switching channel operation fails, we don't change the window size

3.3. Options & preferences:

    Split Digital TV options from analog capture
    New Digital TV options: DVB (BDA) and IPTV (include option: enable/disable IGMP multicast membership management)

Pending issues and known limitations:

    ONLY English version: to be updated once it is ready for merge
    IPTV Channel management: additional options (scan range of IP's and service discovery for multicast) are in progress and will be included in a different PR.
    Access to channel management from options menu: will be also included in a future PR.

**Comment by clsid2** [comment]
If you think the code works good, then importing it is ok. But I don't consider it very important right now.

Are there any sample streams to test that IGMP multicast stuff?

I tried the links from https://github.com/iptv-org/iptv and noticed that opening the URLs directly is broken. The m3u files import correctly after download. So another regression similar to the previous ones.

**Comment by adipose** [comment]
I figure we can close this as wontfix.  The merge is getting old and no one is testing it.  I did my best at porting it, but I really don't know how well it works.  If there is demand we can try to merge it again later, but it's too many changes to bring in without anyone testing it.

**Comment by Eagle3386** [comment]
Sad to see this left out, but then again, I've already switched to VLC since it worked there out of the box as described above.

**Comment by adipose** [comment]
More testing was needed to keep working on it. What works, what doesn't, steps followed. If VLC has you covered maybe we can't expect much help from you, but if you start testing again and give feedback we can reopen this.

**Comment by vBm** [comment]
@adipose Just tried testing but i can't get it to import the list at all. (same list works in android apps and on plex using TellyTV proxy)
Manually adding channels works but when it comes to playing them fails. Again, other players don't have an issue. Tried it also using mpv and i do see some errors decoding audio and packet corruption.

Increasing the "Stream analysis duration" of LAV Splitter from "1000" to "6000" fixed playback for half of the channels i've tried tho. Will try to increase it even more to see if it'll fix with some more.

**Comment by clsid2** [comment]
The internal LAV Filters uses a default value of 2100.

**Comment by vBm** [comment]
Unless the settings for LAV are not reset when making an empty mpc-hc.ini in the same dir then it was set to 1000 for me using the provided build on mega.

**Comment by adipose** [comment]
We can reopen this if we now have a tester.  I probably won't look at it for a little while, though.
